source/common/common/thread.h

```diff
@@ -172,7 +172,24 @@ class AtomicPtr : private AtomicPtrArray<T, 1, alloc_mode> {
 struct MainThread {
   using MainThreadSingleton = InjectableSingleton<MainThread>;
   bool inMainThread() const { return main_thread_id_ == std::this_thread::get_id(); }
-  static void init() { MainThreadSingleton::initialize(new MainThread()); }
+  bool inTestThread() const {
+    return test_thread_id_.has_value() && (test_thread_id_.value() == std::this_thread::get_id());
+  }
+  void registerTestThread() { test_thread_id_ = std::this_thread::get_id(); }
+  void registerMainThread() { main_thread_id_ = std::this_thread::get_id(); }
+  static bool initialized() { return MainThreadSingleton::getExisting() != nullptr; }
+  static void initMainThread() {
+    if (!initialized()) {
```

Comment:
move this non-trivial impl to a .cc file.

Reply:
Makes sense.